templates/JSON/treat_ad.json

```diff
@@ -0,0 +1,118 @@
+{
+    "name": "TREAD_AD_YAML",
+    "type": "Project",
+    "children": [
+      {
+        "name": "Genes",
+        "type": "Folder",
+        "children": [
+          {
+            "name": "MSN",
+            "type": "Folder",
+            "children": [
+              {
+                "name": "Structural_Biology_Core",
+                "type": "Folder",
+                "children": [
+                  "Antibody_Validation_Data",
```

Comment:
Should the long form have each folder expanded into?

```
{"name": "Antibody_Validation_Data"
  "type": "Folder"}
```

Reply:
That's a good point.  Should the YAML template never omit the type then?  Because that's the reason the type didn't make it into these.

Reply:
I was thinking that the yaml could have some "shortcuts", but if we support shortcuts for yaml then we can support them for json. I guess it all comes down whether we want the code to infer the type is it's just a string, or we have code that fills out the expanded format pre creation of synapse resources

Reply:
I'm ok with the code inferring the type in the YAML because the YAML template will never contain files.  In the JSON, you could make the argument that "the leaves" can either be folders (empty folders) or files.

Reply:
Heres what I was thinking.  When reading in the yaml code:  there is a `expand_format` function that essentially expands the shortened form into the long json format.  Due to this, theres no reason why the `yaml` format couldn't actually support Files either.  It could look like:

```
children:
    - My folder
    - name: "my file"
       type: File
```

Technically, theres also no reason why someone couldn't pass in a `json` template that is the shortened version as well, but this long format is good for drafting the "official" template.

Reply:
As far as the input YAML template is concerned, I advocate for the design that I laid out in my original [imCORE draft template](https://github.com/BrunoGrandePhD/sage-sandbox/blob/main/2021-03-10-synapse_formation/drafts/imcore.v1.yaml#L93-L96), which largely is in line with what @thomasyu888 is saying: 

> Top-level resources are assumed to be projects whereas nested resources are assumed to be folders (_i.e._ if `type` isn't given). 
> The key name is assumed to be the resource name on Synapse unless the `name` attribute is given.

That said, I think we should consider making every decision made by SynapseFormation explicit in the output JSON. So, any implicit types would be made explicit in the output, and any names inferred from the key would be stored under the `name:` attribute. Not only would this help with debugging, but it would help users learn what SynapseFormation is doing behind the scenes. This includes adding an `id` field to every resource with the corresponding Synapse or Principal ID that was assigned during creation. 

I envision that this expanded JSON template could serve as input for SynapseFormation, but the client would skip all resources with existing IDs, so it wouldn't change anything. But a user could tweak the JSON (_e.g._ adjust an ACL) and run it through SynapseFormation, which will skip creating entities but will update the adjusted ACL.

Reply:
> Top-level resources are assumed to be projects whereas nested resources are assumed to be folders (i.e. if type isn't given).

As long as it still accepts the keys, I'm indifferent to this.  And I say that because at a personal level this makes it easier for me to write a template, but any newcomer now needs instructions on how to read a template.

> That said, I think we should consider making every decision made by SynapseFormation explicit in the output JSON

Yes, agreed.  Otherwise extracting the structure from a project will not be a symmetric exercise.

> This includes adding an id field to every resource with the corresponding Synapse or Principal ID that was assigned during creation.

Yes, I agree.  Let's plan to include this in future work and scope it properly.

Reply:
> any newcomer now needs instructions on how to read a template.
I think of it as an optional feature. In the documentation, we could use `type` consistently in the examples to make it easier to onboard new users but have a "pro tips" section that points out shortcuts like these for power users.

Reply:
I'm onboard with that.  I think the idea that anyone with some basic YAML knowledge can use our tool is very powerful, and I wouldn't like to use that.